aiida_siesta/workflows/utils/iterate_absclass.py

```diff
@@ -0,0 +1,576 @@
+    The serialization is managed in `_iterate_input_serializer` and `_values_list_serializer`, that can be
+    overridden in case, but this will impose also the change of the `_next_val` method.
+
+    The design of the class also makes this class extensible to do something more than just iterate.
+    The methods cls.return_results, cls.analyze_process, cls._should_proceed can be overridden to support,
+    for instance, a convergence check.
+
+    Following, we display the expanded outline of the workchain in pseudocode to help
+    understand how it works and what can be overidden whithout breaking the code:
+    -------------------------------------------------------------------------------
+    cls.initialize:
+        cls._get_iterator:
+        for key in iterate_over.keys():
+            cls.process_input_and_parse_func(key) # Here all the logic for supported iterations. Can be overridden.
+                                                  # 1) The allowed "key" are defined.
+                                                  # 2) For each key a `_parsing_func` is stored. This function
+                                                  #   implements which input of _process_class needs to be modified
+                                                  #   and how in presence of this "key".
+    while (cls.next_step): # cls._should_proceed and cls._store_next_val are called inside cls.next_step!
+        - cls.run_batch:
+            for i in range(batch_size):
+                if i!= 0:
+                    cls._store_next_val
+                cls._run_process # The current value is under self.current_val and
+                                 # the keys under self.iteration_keys
+        - cls.analyze_batch:
+            for process in batch:
+                cls.analyze_process(process)
+    cls.return_results
+    --------------------------------------------------------------------------------
+    '''
+
+    # THE PROCESS CLASS NEEDS TO BE PROVIDED IN CHILD CLASSES!
+    _process_class = None
+
+    def __init__(self, *args, **kwargs):
+        """Construct the instance. Just needed to implement a check on the passed _process_class"""
+        from aiida.engine import Process  # pylint: disable=cyclic-import
+
+        try:
+            if not issubclass(self._process_class, Process):
+                raise ValueError('no valid Process class defined for `_process_class` attribute')
+        except TypeError:
+            raise ValueError('no valid Process class defined for `_process_class` attribute')
+
+        super().__init__(*args, **kwargs)
+
+    # The inputs of _process_class will always be exposed. This class variable is passed directly
+    # to spec.expose_inputs and can be used, for instance, to exclude some inputs or defining namespaces.
+    _expose_inputs_kwargs = {}  #{'exclude': ('metadata',), "namespace": 'process_class_inputs' }
```

Comment:
I thought you always had to exclude "metadata", otherwise errors will occur, is this not the case?

Reply:
No. If you use namespaces, you resolve the conflict between the class `metadata` and the exposed `metadata`. It is my design of the `SiestaBaseWorkChain` that makes it necessary.

Reply:
Then we could by default exclude metadata and not create a namespace? Otherwise the default `_expose_inputs_kwargs = {}` won't work right? So effectively you are forcing the developer to provide a valid value.

Reply:
We are discussing the topic here: https://github.com/aiidateam/aiida-core/issues/4089
I think the empty `_expose_inputs_kwargs` will just override the class metadata with the `_process_class` metadata port. (Basically you can't pass any metadata to the iterator at this point). Not too bad. I would leave it for the moment, maybe putting a warning in the documentation.

Reply:
Ok :+1: